File: 20241021/haeun/석유시추.js
function solution(land) {
    const dx = [0,1,0,-1];
    const dy = [1,0,-1,0];

    const N = land.length;
    const M = land[0].length;
    const visited = Array.from({length: N}, ()=> new Array(M).fill(0));
    const petroleum = [0];
    
    let maxDrilled = 0;
    let loafNumber = 1;
    
    // 덩어리마다 번호 매기기
    // 번호에 해당하는 petroleum 인덱스에 덩어리 크기 저장
    for(let i = 0; i < N; i++){
        for(let j = 0; j < M; j++){
            if(visited[i][j] === 0 && land[i][j] === 1){
                visited[i][j] = loafNumber;
                const amount = bfs(i,j);
                petroleum.push(amount);
                loafNumber++;
            }
        }
    }
    
    const petroleumLen = petroleum.length;
    
    // 열마다 덩어리가 몇개 닿는지 검사
    for(let j = 0; j < M; j++){
        const loafVisited =  new Array(petroleumLen).fill(0);
        let count = 0;
        for(let i = 0; i < N; i++){
            const targetLoaf = visited[i][j];
            if(land[i][j] === 1 && loafVisited[targetLoaf] === 0){
                loafVisited[targetLoaf] = 1;
                count += petroleum[targetLoaf];
                
            }
        }
        maxDrilled = Math.max(maxDrilled, count);
        
    }
    
    
    function bfs(r, c){
    const queue = [[r,c,1]];
    let head = 0;
    let count = 1;
    
    while(queue.length > head){
        const [x,y] = queue[head++];
        for(let i = 0; i < 4; i++){
            
            const [nx, ny] = [x+dx[i], y + dy[i]];
        
            if(nx < 0 || nx >= N || ny < 0 || ny >= M || visited[nx][ny] || !land[nx][ny]) continue;
            visited[nx][ny] = loafNumber;
            count++;
            queue.push([nx, ny]);
        }
        
    }
        return count;
}
    
    return maxDrilled;
}

File: 20241022/donggyu/석유_시추.java
import java.io.*;
import java.util.*;
class Solution {
    static int[][] deltas = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
    static int[] oilBlock;
    
    public int solution(int[][] land) {
        int answer = 0;
        boolean [][] visited = new boolean[land.length][land[0].length];
        oilBlock = new int[land[0].length];
		for(int i=0;i<land.length;i++) {
			for(int j=0;j<land[0].length;j++) {
                if(land[i][j] != 0 && !visited[i][j]){
                    bfs(i, j, land, visited);				
                }
			}
		}
		
        for(int i=0;i<land[0].length;i++) {
        	answer = Math.max(answer, oilBlock[i]);
        }
        
        System.out.println(answer);
        return answer;
    }
    
    static void bfs(int x, int y, int[][] land, boolean[][] visited) {
		Queue<int[]> q = new ArrayDeque<>();
		q.add(new int[]{x, y});
        visited[x][y] = true;
		int size = 1;
		
		Set<Integer> set = new HashSet<>();
		while(!q.isEmpty()) {
			int[] now = q.poll();
			int nowX = now[0], nowY = now[1];
			set.add(nowY);
			
			for(int[] d : deltas) {
				int nx = nowX+d[0];
				int ny = nowY+d[1];
				if(nx<0 || ny<0 || nx>=land.length||ny>=land[0].length) continue;
				if(visited[nx][ny]) continue;
                if(land[nx][ny]==0) continue;
				q.add(new int[] {nx, ny});
                visited[nx][ny] = true;
                size++;
			}
		}
		
		for (int oilY : set) {
			oilBlock[oilY] += size;
		}
	}
}